src/porepy/viz/exporter.py

```diff
@@ -520,546 +1064,789 @@ def _export_grid(
+                        for i in range(2)
+                    ]
+                    # Group first and second nodes, with alternating order of rows.
+                    # By this, each cell is respresented by two rows. The first and second
+                    # rows contain first and second nodes of faces. And each column stands
+                    # for one face.
+                    cfn = np.ravel(cfn_indices, order="F").reshape(n, -1).T
+
+                    # Sort faces for each cell such that they form a chain. Use a function
+                    # compiled with Numba. This step is the bottleneck of this routine.
+                    cfn = self._sort_point_pairs_numba(cfn).astype(int)
+
+                    # For each cell pick the sorted nodes such that they form a chain
+                    # and thereby define the connectivity, i.e., skip every second row.
+                    cn_indices = cfn[::2, :]
+
+                # Add offset to account for previous grids, and store
+                cell_to_nodes[cell_type] = np.vstack(
+                    (cell_to_nodes[cell_type], cn_indices + nodes_offset)
+                )
 
-            pts_pos += g.num_nodes
+            # Update offsets
+            nodes_offset += g.num_nodes
+            cell_offset += g.num_cells
 
-        # construct the meshio data structure
-        num_block = len(cell_to_nodes)
-        meshio_cells = np.empty(num_block, dtype=object)
-        meshio_cell_id = np.empty(num_block, dtype=object)
+        # Construct the meshio data structure
+        num_blocks = len(cell_to_nodes)
+        meshio_cells = np.empty(num_blocks, dtype=object)
+        meshio_cell_id = np.empty(num_blocks, dtype=object)
 
+        # For each cell_type store the connectivity pattern cell_to_nodes for
+        # the corresponding cells with ids from cell_id.
         for block, (cell_type, cell_block) in enumerate(cell_to_nodes.items()):
-            # remove the number of nodes associated to the polygon since meshio requires only
-            # the keyword "polygon" for polygons
-            cell_type_ = "polygon" if "polygon" in cell_type else cell_type
-            meshio_cells[block] = meshio.CellBlock(cell_type_, cell_block.astype(int))
+
+            # Meshio requires the keyword "polgon" for general polygons.
+            # Thus, remove the number of nodes associated to polygons.
+            cell_type_meshio_format = "polygon" if "polygon" in cell_type else cell_type
+            meshio_cells[block] = meshio.CellBlock(
+                cell_type_meshio_format, cell_block.astype(int)
+            )
             meshio_cell_id[block] = np.array(cell_id[cell_type])
 
-        return meshio_pts, meshio_cells, meshio_cell_id
+        # Return final meshio data: points, cell (connectivity), cell ids
+        return Meshio_Geom(meshio_pts, meshio_cells, meshio_cell_id)
+
+    def _sort_point_pairs_numba(self, lines: np.ndarray) -> np.ndarray:
```

Comment:
No chance of avoiding code duplication? If no, could we colocate the two functions? It seems self is not accessed.

Reply:
@keileg , what do you think, as I believe you have been using the original routine. `_sort_point_pairs_numba` as defined here has the simplification that it assumes circular polygons and on the other hand allows for a list of them instead of only considering a single one. I stripped the routine down to its essentials to allow for numba - this was the main motivation.  Moving it to utils could be indeed an idea.

Reply:
Moving to utils.point_pairs makes sense to me. As @jwboth says, the two functions make different assumptions, thus a unified function will get some keywords, and then an if-else that essentially calls one of the two methods. That may or may not be an improvement, and the perspective may be different from the user and maintainer side.

I suggest we first move it, and then see what it looks like.

Reply:
Done as @keileg suggested.